pandas/indexes/base.py

```diff
@@ -1403,46 +1407,31 @@ def _ensure_compat_append(self, other):
             to_concat.append(other)
 
         for obj in to_concat:
-            if (isinstance(obj, Index) and obj.name != name and
-                    obj.name is not None):
-                name = None
-                break
+            if not isinstance(obj, Index):
+                raise TypeError('all inputs must be Index')
```

Comment:
``_ensure_index`` should handle this (it *might* think its a ``MultiIndex`` actually), so might need to disambiguate